source/server/options_impl.cc

```diff
@@ -93,24 +93,36 @@ OptionsImpl::OptionsImpl(int argc, char** argv, const HotRestartVersionCb& hot_r
                                              false, ENVOY_DEFAULT_MAX_OBJ_NAME_LENGTH, "uint64_t",
                                              cmd);
 
+  cmd.setExceptionHandling(false);
   try {
     cmd.parse(argc, argv);
   } catch (TCLAP::ArgException& e) {
-    std::cerr << "error: " << e.error() << std::endl;
-    exit(1);
+    try {
+      cmd.getOutput()->failure(cmd, e);
+    } catch (const TCLAP::ExitException&) {
+      // failure() has already written an informative message to stderr, so all that's left to do
+      // is throw our own exception with the original message.
+      throw MalformedArgvException(e.what());
+    }
+  } catch (const TCLAP::ExitException& e) {
+    // parse() throws an ExitException with status 0 after printing the output for --help and
+    // --version.
+    throw NoServingException();
   }
 
   if (max_obj_name_len.getValue() < 60) {
-    std::cerr << "error: the 'max-obj-name-len' value specified (" << max_obj_name_len.getValue()
-              << ") is less than the minimum value of 60" << std::endl;
-    exit(1);
+    const std::string message = fmt::format(
+        "error: the 'max-obj-name-len' value specified ({}) is less than the minimum value of 60",
+        max_obj_name_len.getValue());
+    std::cerr << message << std::endl;
+    throw MalformedArgvException(message);
   }
 
   if (hot_restart_version_option.getValue()) {
     std::cerr << hot_restart_version_cb(max_stats.getValue(),
                                         max_obj_name_len.getValue() +
                                             Stats::RawStatData::maxStatSuffixLength());
-    exit(0);
+    throw NoServingException();
```

Comment:
I dislike throwing exceptions for non-error cases.  On the other hand, the alternative is setting a member boolean and adding something to the interface to check for it, and I don't know if that's any cleaner.  So I don't dislike it enough to say "no", but I'm curious if anyone else has thoughts opinions on this?

Reply:
(@htuch and I were talking about this earlier.)

FWIW, I agree in general it's distasteful to use exceptions for success cases. Here, we're talking about a set of arguments that can't be used to construct a usable OptionsImpl, so it seems reasonable to treat it as an error in that sense, even though the exit status will be zero. ("Usable" in the sense that you couldn't start a server with it -- I don't think it's a good idea to construct an OptionsImpl that can't be passed into the rest of the startup flow.) I'm definitely open to counter-proposals, though.

Reply:
Yeah I think the only other solution is @ggreenway proposal to add state/method and not throw exception. I agree in general this pattern smells a little bad to me, but in this case, I don't have a strong opinion. I'm fine either way.

Reply:
My take is basically the same as @rlazarus, that `NoServingException` is an exceptional condition; normally we expect to spin up the server with a valid set of options, we don't have that here. It's just a different flavor of error.